### PR TITLE
Final out csv

### PR DESCRIPTION
**Pull request by nick-pestell**
This PR adapts `btb-phylo` code to work with the `FinalOut.csv` files for collecting sample metadata rather than a larger csv containing redundant metadata. 

**Main additions**:
- `update_summary.py`: This code updates a core *summary csv* file which is essentially a concatenation of `FinalOut.csv` for all batches with the addition of a column for the submission number of each sample. In detail it ... ***i)*** scans s3 for all available `FinalOut.csv` files (`get_finalout_s3_keys()`); ***ii)*** finds the "new keys", relating to new samples by comparing to those already in the *summary csv*; ***iii)*** downloads the new `FinalOut.csv` files (`finalout_csv_to_df()`), adds a submission number column (`add_submission_col()`); ***iv)*** appends to the *summary csv* (`append_df_summary()`); and ***v)*** pushes the summary csv to s3 (`df_to_s3()`)

This will take a few minutes to run when first run, because it populates the *summary csv* from scratch, If this is used in production it will download the partially populated *summary csv* from s3 and only update with new samples, therefore it will be much quicker. 

- `FinalOut.csv` contains a `ResultLoc` column for each sample, which is an s3_uri. New functions, `extract_s3_bucket()` and `extract_s3_key()` in `btb_phylo.py` are used to extract the bucket and key from the s3_uri for use with `boto3` library in order to download consensus files.
- New unit tests have been added for testing `test_extract_submission_no()` and `test_append_summary_df()` for testing two these two functions from `update_summary.py`. Also a new unit test `test_extract_s3_bucket()`, `test_extract_s3_key()` and `test_match_s3_uri()` test new functions in `btb_phylo.py`

### CODE REVIEW

File: btb_phylo.py [hunk 1]
```diff
@@ -65,18 +54,18 @@ def get_indexes_to_remove(df, parameter):
...
-    for submission_no in df.submission.unique():
-        parameter_max = df.loc[df["submission"]==submission_no][parameter].max()
-        if len(df.loc[(df["submission"]==submission_no) & (df[parameter] == parameter_max)]) > 1:
-            warnings.warn(f"Submision {submission_no} is duplicated and has the same "
-                          f"{parameter} value\nSkipping submision {submission_no}")
+    for submission_no in df.Submission.unique():
+        parameter_max = df.loc[df["Submission"]==submission_no][parameter].max()
+        if len(df.loc[(df["Submission"]==submission_no) & (df[parameter] == parameter_max)]) > 1:
+            warnings.warn(f"Submision {submission_no} is duplicated and duplicates have "
+                          f"the same {parameter} value\nSkipping submision {submission_no}")
```
Comment (ellisrichardj) [inline btb_phylo.py:61]: Suggest that one is retained?

File: unit_tests.py [hunk 2]
```diff
@@ -192,11 +196,195 @@ def test_build_multi_fasta(self):
...
+                      "12-3456-89-1L",
+                      "12-3456-89-L1",
+                      "A-12-3456-89",
+                      "12-3456-89-1",
+                      "12-3456-89-L",
+                      "12-3456-89",
+                      "12345678",
+                      "ABCDEFGH",
+                      ""]
+        test_output = ["12-34567-89",
```
Comment (ellisrichardj) [inline unit_tests.py:290]: Submission numbers all have an `AF-` prefix in LIMS, but is this what's required for matching with other data sources?

**Review by ellisrichardj — COMMENTED** [review]
Great to see these amendments and additions to the code.  The implementation of just appending a new file looks good, but in my experience the `aws s3 ls` is slower than I would have expected, but not sure what other options there are.  There are a few comments in particular sections of the code too.

File: btb_phylo.py (hunk 1 above)
Comment (nick-pestell) [reply btb_phylo.py:61]: Yes, I think you're right we should retain one. I was supprised to see at least 100 samples with the same submission number and identical `pcMapped` value. Not sure what the policy should be for choosing which one to retain, perhaps go on a second parameter? `Ncount` might be better than pcMapped for choosing samples too? Will look into this tomorrow.

File: btb_phylo.py [hunk 3]
```diff
@@ -65,18 +54,18 @@ def get_indexes_to_remove(df, parameter):
             indexes (pandas index object): indexes to remove from dataframe
     """
     indexes = pd.Index([])
-    for submission_no in df.submission.unique():
-        parameter_max = df.loc[df["submission"]==submission_no][parameter].max()
-        if len(df.loc[(df["submission"]==submission_no) & (df[parameter] == parameter_max)]) > 1:
-            warnings.warn(f"Submision {submission_no} is duplicated and has the same "
-                          f"{parameter} value\nSkipping submision {submission_no}")
+    for submission_no in df.Submission.unique():
+        parameter_max = df.loc[df["Submission"]==submission_no][parameter].max()
```
[the remark this answers is not in the document]
Comment (nick-pestell) [reply btb_phylo.py:58]: Yeah good point, perhaps I can add min/max as an argument to the function.

File: unit_tests.py (hunk 2 above)
Comment (nick-pestell) [reply unit_tests.py:290]: Not too sure what you mean, I was just using the submission number (without AF) to choose which samples to include. The snp matrix has the original sample names, i.e. AF-....

File: btb_phylo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nick-pestell) [reply btb_phylo.py]: I've now adapted the function so that the user can specify whether to use a min or max method along with the parameter they choose.

**Review by ellisrichardj — APPROVED** [review]
Thanks for implementing the additional changes

**Pull request merged**